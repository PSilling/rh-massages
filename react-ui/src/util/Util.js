/**
 * Utility class containing project-wide utility functions.
 */

// module imports
import moment from "moment";
import { NotificationManager } from "react-notifications";

// util imports
import _t from "./Translations";

const Util = function Util() {};

/**
 * Creates a new notification message. Supports types info, success, warning and error.
 * @param type            notification type
 * @param message         notification message
 * @param title           notification title
 * @param timeout         notification timeout in milliseconds
 * @param callback        callback fired on notification click
 * @param priority        whether the notification should be displayed as a priority one on top
 */
Util.notify = (type, message, title, timeout = 2000, callback = () => {}, priority = true) => {
  switch (type) {
    case "info":
      NotificationManager.info(message, title, timeout, callback, priority);
      break;
    case "success":
      NotificationManager.success(message, title, timeout, callback, priority);
      break;
    case "warning":
      NotificationManager.warning(message, title, timeout, callback, priority);
      break;
    case "error":
      NotificationManager.error(message, title, timeout, callback, priority);
      break;
    default:
      NotificationManager.info(message, title, timeout, callback, priority);
      break;
  }
};

/**
 * Checks whether an Object is null, undefined or an empty string.
 * @param  object Object to check
 * @return true if empty, false otherwise
 */
Util.isEmpty = function isEmpty(object) {
  return object === null || typeof object === "undefined" || object === "";
};

/**
 * Clears all active intervals in the client.
 */
Util.clearAllIntervals = () => {
  const maxIntervalId = setTimeout(() => {
    for (let i = 0; i < maxIntervalId; i++) {
      clearInterval(i);
    }
  }, 0);
};

/**
 * Generates a Google Calendar link for a given massage event.
 * @param massage the Massage to be added to the calendar
 */
Util.getEventLink = massage =>
  "https://www.google.com/calendar/render?action=TEMPLATE" +
  `&text=${_t.translate("Massage")}` +
  `&dates=${moment
    .utc(massage.date)
    .format("YYYYMMDDTHHmmssZ")
    .replace("+00:00", "Z")}` +
  `/${moment
    .utc(massage.ending)
    .format("YYYYMMDDTHHmmssZ")
    .replace("+00:00", "Z")}` +
  `&location=Red%20Hat%20Czech%20${massage.facility.name}` +
  `&details=${_t.translate("Masseur/Masseuse")}: ${Util.getContactInfo(massage.masseuse)}`;

/**
 * Creates a new contact info String from a given Client.
 * @param client Client to get contact info for
 */
Util.getContactInfo = client => `${client.name} ${client.surname} (${client.email})`;

/**
 * Searches an array for a given item ID.
 * @param array array to search in
 * @param id ID to search for
 * @param idKey attribute name of the ID
 */
Util.findInArrayById = (array, id, idKey = "id") => {
  for (let i = 0; i < array.length; i++) {
    if (array[i][idKey] === id) {
      return i;
    }
  }
  return -1;
};

/**
 * Searches a massage event array for a given massage ID.
 * @param array array to search in
 * @param id id to search for
 */
Util.findInArrayByMassageId = (array, id) => {
  for (let i = 0; i < array.length; i++) {
    if (array[i].massage.id === id) {
      return i;
    }
  }
  return -1;
};

/**
 * Returns an array of tooltip targets for unique tooltip ID use.
 * @param  count number of tooltip targets to generate
 * @return       array of generated tooltip targets
 */
Util.getTooltipTargets = count => {
  const targets = [];
  for (let i = 0; i < count; i++) {
    targets.push(`Tooltip${Util.tooltipCount++}`);
  }
  return targets;
};

/**
 * Check whether a given massage would be over the month time limit or not.
 * @param  massageMinutes number of currently used Massage times per each month in minutes
 * @param  massage massage to be evaluated
 * @return true if the massage is over the limit, false otherwise
 */
Util.isOverTimeLimit = (massageMinutes, massage) => {
  let minutesTotal = moment(massage.ending).diff(moment(massage.date), "minutes");

  const month = moment(massage.date).format("MM-YYYY");
  if (Object.hasOwnProperty.call(massageMinutes, month)) {
    minutesTotal += massageMinutes[month];
  }

  return minutesTotal > Util.MAX_MASSAGE_MINS;
};

/** url of /facilities endpoint */
Util.FACILITIES_URL = process.env.REACT_APP_FACILITIES_URL || "api/facilities/";
/** url of /massages endpoint */
Util.MASSAGES_URL = process.env.REACT_APP_MASSAGES_URL || "api/massages/";
/** url of /clients endpoint */
Util.CLIENTS_URL = process.env.REACT_APP_CLIENTS_URL || "api/clients/";
/** url of /websockets authentication endpoint */
Util.WEBSOCKETS_URL = process.env.REACT_APP_WEBSOCKETS_URL || "api/websockets";
/** GitHub project url */
Util.GITHUB_URL = process.env.REACT_APP_GITHUB_URL || "https://github.com/";
/** refresh time for authorization tokens in milliseconds */
Util.REFRESH_MIN_TIME = parseInt(process.env.REACT_APP_MIN_REFRESH_TIME, 10) || 150;
/** cancellation limit before the start of a Massage in minutes */
Util.CANCELLATION_LIMIT = parseInt(process.env.REACT_APP_CANCELLATION_LIMIT, 10) || 30;
/** maximum minute time of Massages per Client */
Util.MAX_MASSAGE_MINS = parseInt(process.env.REACT_APP_MASSAGE_TIME_LIMIT, 10) || 120;
/** url of /websockets subscription endpoint with a possible proxy and Websocket protocol */
Util.WEBSOCKET_PROTOCOL_URL = process.env.REACT_APP_WEBSOCKET_PROTOCOL_URL || "wss://api/websockets";
/** maximum amount of WebSocket handshake request retries; wait time increases exponentionally */
Util.WEBSOCKET_RETRY_COUNT = parseInt(process.env.REACT_APP_WEBSOCKET_RETRY_COUNT, 10) || 3;
/** maximum wait time for WebSocket actions */
Util.WEBSOCKET_TIMEOUT_LIMIT = process.env.REACT_APP_WEBSOCKET_TIMEOUT_LIMIT || 250;
/** display color of success events */
Util.SUCCESS_COLOR = process.env.REACT_APP_SUCCESS_COLOR || "#2fad2f";
/** display color of warning events */
Util.WARNING_COLOR = process.env.REACT_APP_WARNING_COLOR || "#ee9d2a";
/** display color of error events */
Util.ERROR_COLOR = process.env.REACT_APP_ERROR_COLOR || "#d10a14";

/** current number of tooltips (needed for correct ID placement) */
Util.tooltipCount = 1;

export default Util;
