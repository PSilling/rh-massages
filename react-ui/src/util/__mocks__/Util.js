const get = jest.fn((url, update) => {
  update([]);
});
const post = jest.fn((url, data, update) => {
  update();
});
const put = jest.fn((url, data, update) => {
  update();
});
const notify = jest.fn();
const clearAllIntervals = jest.fn();
const addToCalendar = jest.fn();

const isEmpty = jest.fn(object => object === null || typeof object === "undefined" || object === "");

const findInArrayById = jest.fn((array, id) => {
  for (let i = 0; i < array.length; i++) {
    if (array[i].id === id) {
      return i;
    }
  }
  return -1;
});

module.exports = {
  get,
  post,
  put,
  notify,
  clearAllIntervals,
  addToCalendar,
  isEmpty,
  findInArrayById
};
