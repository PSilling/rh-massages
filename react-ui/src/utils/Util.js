import moment from 'moment';

import _t from './Translations';
import { NotificationManager } from 'react-notifications';
import Auth from './Auth';

var Util = function() { };

/**
 * Checks whether an object is null or undefined
 * @param  object to check
 * @return boolean
 */
Util.isEmpty = function(object) {
  return (object === null) || (typeof object === 'undefined')
}

/**
 * Creates a new notification.
 * @param type            notification type
 * @param message         notification message
 * @param title           notification title
 */
Util.notify = (type, message, title) => {
  switch (type) {
    case 'info':
      NotificationManager.info(message, title, 2000);
      break;
    case 'success':
      NotificationManager.success(message, title, 2000, null, true);
      break;
    case 'warning':
      NotificationManager.warning(message, title, 2000);
      break;
    case 'error':
      NotificationManager.error(message, title, 2000, null, true);
      break;
    default:
      NotificationManager.info(message, title, 2000);
      break;
  };
}

/**
 * Fetches data from a given endpoint.
 * @param url             defined endpoint
 * @param update          callback function to update the resources
 */
Util.get = (url, update) => {
  Auth.keycloak.updateToken(Util.REFRESH_MIN_TIME).success(function() {
    fetch(url, {
      method: 'get',
      headers: {
        'Authorization': 'Bearer ' + Auth.getToken()
      }
    }).then(function(response) {
      if (response.ok) {
        return response.json();
      } else {
        Util.notify("error", _t.translate('Your request has ended unsuccessfully.'),
          _t.translate('An error occured!'));
      }
    }).then(function(json) {
      console.log(json);
      update(json);
    });
  }).error(function() {
    console.log('Failed to refresh the token!');
    Auth.keycloak.login();
  });
}

/**
 * Creates a new element at a given endpoint.
 * @param url             defined endpoint
 * @param data            data to send
 * @param update          callback function to update the resources
 */
Util.post = (url, data, update) => {
  Auth.keycloak.updateToken(Util.REFRESH_MIN_TIME).success(function() {
    console.log(data);
    fetch(url, {
      method: 'post',
      credentials: 'same-origin',
      headers: {
        "Authorization" : "bearer " + Auth.getToken(),
        "Content-Type" : "application/json"
      },
      body: JSON.stringify(data)
    }).then(function(response) {
      if (response.ok) {
        Util.notify("success", "", _t.translate('Your request has been successful.'));
        update();
      } else {
        Util.notify("error", _t.translate('Your request has ended unsuccessfully.'),
          _t.translate('An error occured!'));
      }
    });
  }).error(function() {
    console.log('Failed to refresh the token!');
    Auth.keycloak.login();
  });
}

/**
 * Edits an element at a given endpoint.
 * @param url             defined endpoint
 * @param data            data to send
 * @param update          callback function to update the resources
 */
Util.put = (url, data, update) => {
  Auth.keycloak.updateToken(Util.REFRESH_MIN_TIME).success(function() {
    console.log(data);
    fetch(url, {
      method: 'put',
      credentials: 'same-origin',
      headers: {
        "Authorization" : "bearer " + Auth.getToken(),
        "Content-Type" : "application/json"
      },
      body: JSON.stringify(data)
    }).then(function(response) {
      if (response.ok) {
        Util.notify("success", "", _t.translate('Your request has been successful.'));
        update();
      } else {
        Util.notify("error", _t.translate('Your request has ended unsuccessfully.'),
          _t.translate('An error occured!'));
      }
    });
  }).error(function() {
    console.log('Failed to refresh the token!');
    Auth.keycloak.login();
  });
}

/**
 * Deletes an element at a given endpoint.
 * @param url             defined endpoint
 * @param update          callback function to update the resources
 */
Util.delete = (url, update) => {
  Auth.keycloak.updateToken(Util.REFRESH_MIN_TIME).success(function() {
    fetch(url, {
      method: 'delete',
      credentials: 'same-origin',
      headers: {
        "Authorization" : "bearer " + Auth.getToken()
      }
    }).then(function(response) {
      if (response.ok) {
        Util.notify("success", "", _t.translate('Your request has been successful.'));
        update();
      } else {
        Util.notify("error", _t.translate('Your request has ended unsuccessfully.'),
          _t.translate('An error occured!'));
      }
    });
  }).error(function() {
    console.log('Failed to refresh the token!');
    Auth.keycloak.login();
  });
}

/**
 * Stops all active intervals.
 */
Util.clearAllIntervals = () => {
  var maxIntervalId = setTimeout(function() {
    for (var i = 0; i < maxIntervalId; i++) {
        clearInterval(i);
    }
  }, 0);
}

/**
 * Adds a massage event to Google Calendar (opens a new tab).
 *
 * @param massage the massage to be added to the calendar
 */
Util.addToCalendar = (massage) => {
  var url = "https://www.google.com/calendar/render?action=TEMPLATE";
  url += "&text=" + _t.translate('Massage in facility') + ' ' + massage.facility.name;
  url += "&dates=" + moment.utc(massage.date).format("YYYYMMDDTHHmmssZ").replace("+00:00", "Z");
  url += "/" + moment.utc(massage.date).add(1, 'h').format("YYYYMMDDTHHmmssZ").replace("+00:00", "Z");
  url += "&details=" + _t.translate('Masseuse') + ' ' + massage.masseuse;

  window.open(url,"_blank");
}

Util.FACILITIES_URL = "/api/facilities/";
Util.MASSAGES_URL = "/api/massages/";
Util.LOGOUT_URL = "/api/logout/";
Util.REFRESH_MIN_TIME = 150;
Util.AUTO_REFRESH_TIME = 10000;

export default Util
