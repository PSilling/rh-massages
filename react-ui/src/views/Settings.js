// react imports
import React, { Component } from "react";

// component imports
import InfoAlert from "../components/util/InfoAlert";
import "../styles/components/loader.css";

// util imports
import Auth from "../util/Auth";
import _t from "../util/Translations";
import Util from "../util/Util";

/**
 * View containing portal and user settings.
 */
class Settings extends Component {
  state = { notify: false, loading: true };

  alertMessage =
    _t.translate("On this page you can manage your local user settings. ") +
    _t.translate("You can also use this page to access our repository on GitHub.");

  componentDidMount() {
    Util.clearAllIntervals();

    this.getSettings();
    setInterval(() => {
      this.getSettings();
    }, Util.AUTO_REFRESH_TIME * 600);
  }

  getSettings = () => {
    Util.get(`${Util.CLIENTS_URL}my/subscribed`, json => {
      this.setState({ notify: json, loading: false });
    });
  };

  changeNotify = event => {
    Auth.subscribed = event.target.checked;
    Util.put(
      Util.CLIENTS_URL,
      Auth.getClient(),
      () => {
        this.setState({ notify: Auth.subscribed });
      },
      false
    );
  };

  closeAlert = () => {
    localStorage.setItem("closeSettingsAlert", true);
    this.setState(prevState => ({ loading: prevState.loading }));
  };

  render() {
    return (
      <div>
        {!localStorage.getItem("closeSettingsAlert") ? (
          <InfoAlert onClose={this.closeAlert}>{this.alertMessage}</InfoAlert>
        ) : (
          ""
        )}
        <h1>
          {this.state.loading ? <div className="loader pull-right" /> : ""}
          {_t.translate("Settings")}
        </h1>
        <hr />
        <h3>{_t.translate("User settings")}</h3>
        <div className="row">
          <div className="col-md-12">
            <div className="form-group">
              <label className="checkbox-inline" htmlFor="notifyInput">
                <input id="notifyInput" type="checkbox" onChange={this.changeNotify} checked={this.state.notify} />
                {_t.translate("I want to recieve information about ")}
                <abbr
                  title={_t.translate("Scheduling of new and cancellation of assigned massages")}
                  style={{ cursor: "pointer" }}
                >
                  {_t.translate("massage changes")}
                </abbr>
              </label>
            </div>
          </div>
        </div>

        <h3>{_t.translate("About")}</h3>
        <div className="row">
          <div className="col-md-12">
            {_t.translate("Visit our ")}
            <a href={Util.GITHUB_URL} target="_blank" rel="noreferrer noopener">
              GitHub
            </a>
            &nbsp;
            <del>
              {_t.translate(" and our ")}
              <a href={`${Util.GITHUB_URL}wiki`} target="_blank" rel="noreferrer noopener">
                wiki
              </a>
            </del>
            ! (WIP)
          </div>
        </div>
        <div className="row">
          <div className="col-md-12">
            {_t.translate("Report issues ")}
            <a href={`${Util.GITHUB_URL}issues`} target="_blank" rel="noreferrer noopener">
              {_t.translate("here")}
            </a>
            .
          </div>
        </div>
      </div>
    );
  }
}

export default Settings;
