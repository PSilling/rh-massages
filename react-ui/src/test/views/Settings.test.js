// react imports
import React from "react";
import TestRenderer from "react-test-renderer";

// test imports
import Settings from "../../views/Settings";
import Util from "../../util/Util";

// test mocks
jest.mock("../../util/Auth");
jest.mock("../../util/Util");

afterAll(() => {
  jest.resetAllMocks();
});

test("renders content correctly", () => {
  const testRenderer = TestRenderer.create(<Settings />);
  const treeJSON = testRenderer.toJSON();

  expect(treeJSON).toMatchSnapshot();
});

test("properly changes state variables", () => {
  Util.get = jest.fn((url, update) => {
    update(true);
  });
  const testRenderer = TestRenderer.create(<Settings />);
  const testInstance = testRenderer.root;

  testInstance.instance.setState({ loading: true, notify: false });
  testInstance.instance.getSettings();
  expect(testInstance.instance.state.notify).toEqual(true);
  expect(testInstance.instance.state.loading).toBe(false);
  testInstance.findAllByProps({ className: "row" });

  const notifyInput = testInstance.findByProps({
    onChange: testInstance.instance.changeNotify
  });

  expect(notifyInput.props.checked).toBe(true);
  notifyInput.props.onChange({ target: { checked: false } });
  expect(notifyInput.props.checked).toBe(false);
  expect(Util.put).toHaveBeenCalledTimes(1);
});
