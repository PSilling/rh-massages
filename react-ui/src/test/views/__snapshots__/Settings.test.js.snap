// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content correctly 1`] = `
<div>
  <div
    className="alert alert-info alert-dismissible"
    style={
      Object {
        "marginBottom": "15px",
      }
    }
  >
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      title="Zavřít"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    Na této stránce můžete spravovat svoje lokální uživatelské nastavení. Také lze tuto stránku použít k přístupu k našemu repozitáři na GitHubu.
  </div>
  <h1>
    
    Nastavení
  </h1>
  <hr />
  <h3>
    Uživatelská nastavení
  </h3>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      <div
        className="form-group"
      >
        <label
          className="checkbox-inline"
          htmlFor="notifyInput"
        >
          <input
            checked={Array []}
            id="notifyInput"
            onChange={[Function]}
            type="checkbox"
          />
          Chci dostávat informace o 
          <abbr
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Rozpis nových a zrušení připsaných masáží"
          >
            změnách masáží
          </abbr>
        </label>
      </div>
    </div>
  </div>
  <h3>
    O portálu
  </h3>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      Navštivte náš 
      <a
        href={undefined}
        rel="noreferrer noopener"
        target="_blank"
      >
        GitHub
      </a>
       
      <del>
        a naši 
        <a
          href="undefinedwiki"
          rel="noreferrer noopener"
          target="_blank"
        >
          wiki
        </a>
      </del>
      ! (WIP)
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      Hlaste chyby 
      <a
        href="undefinedissues"
        rel="noreferrer noopener"
        target="_blank"
      >
        zde
      </a>
      .
    </div>
  </div>
</div>
`;
